filegenerator/file_generator.go

```diff
@@ -0,0 +1,361 @@
+)
+
+// BenchPackages is a mapping of package names to test names, which in turn map
+// to their historical data.
+type BenchPackages map[string]BenchTestMap
+
+// BenchTestMap is a mapping of test names to dates, which in turn map to the
+// results of the benchmark tests for that given date.
+type BenchTestMap map[string]BenchResults
+
+// BenchResults map dates to the benchmark test data for that date.
+type BenchResults map[string]BenchStats
+
+// BenchStats is a struct containing the various metrics obtained from running
+// a benchmark test with the -benchmem parameter. The variable names are kept
+// short to keep the JSON files small.
+type BenchStats struct {
+	N int     // ns/op
+	A int     // alllocs/op
+	B int     // B/op
+	M float64 // MB/s
+}
+
+// GeometricMeanData is a struct that contains the means of the benchmark
+// stats for all the tests in a given package.
+type GeometricMeanData struct {
+	NMean float64 // Mean of the ns/op result from all tests in a package.
+	AMean float64 // Mean of the allocs/op result from all tests in a package.
+	BMean float64 // Mean of the B/op result from all tests in a package.
+	MMean float64 // Mean of the MB/s result from all tests in a package.
+	Date  string  // The date that the mean of all tests was collected.
+}
+
+func isValidTestLine(line string) bool {
+	matched, err := regexp.MatchString("FAIL", line)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// This matches a line from Stdout where a benchmark test passes.
+	return len(line) >= 9 && line[0:9] == "Benchmark" && !matched
+}
+
+func parseTestResultLine(line string) (string, BenchStats) {
+	testName := strings.Trim(strings.Split(line, "\t")[0], " ")
+	nsopr := regexp.MustCompile("[0-9]+ ns/op")
+	allocpr := regexp.MustCompile("[0-9]+ allocs/op")
+	bopr := regexp.MustCompile("[0-9]+ B/op")
+	mbpsr := regexp.MustCompile("[0-9]+\\.?[0-9]+ MB/s")
+	var nsopi, allocpi, bopi int
+	var mbpsf float64
+	if nsops := nsopr.FindString(line); nsops != "" {
+		nsopi, _ = strconv.Atoi(strings.Split(nsops, " ")[0])
+	}
+	if allocps := allocpr.FindString(line); allocps != "" {
+		allocpi, _ = strconv.Atoi(strings.Split(allocps, " ")[0])
+	}
+	if bops := bopr.FindString(line); bops != "" {
+		bopi, _ = strconv.Atoi(strings.Split(bops, " ")[0])
+	}
+	if mbpss := mbpsr.FindString(line); mbpss != "" {
+		mbpsf, _ = strconv.ParseFloat(strings.Split(mbpss, " ")[0], 64)
+	}
+	return testName, BenchStats{N: nsopi, A: allocpi, B: bopi, M: mbpsf}
+}
+
+func isValidDataDir(dataDirName string) bool {
+	matched, _ := regexp.MatchString("[0-9]{2}-[0-9]{2}-[0-9]{4}", dataDirName)
+	return matched
+}
+
+// RenderHistoricalBenchmarkResults takes the folders in benchSamples, parses them, and returns a
+// data structure called BenchPackages which represent the data from the folders.
+func RenderHistoricalBenchmarkResults(dirs []string) BenchPackages {
+	dirToTestNames := make(BenchPackages)
+	benchSamples := os.Getenv(benchSamplesEnv)
```

Comment:
Should you check if `benchSamples` is `""` and error accordingly? It seems like you'd possibly weird errors below (e.g. `ReadDir`) if `benchSampleEnv` wasn't set.